main/src/main/scala/sbt/Defaults.scala

```diff
@@ -719,7 +719,7 @@ object Defaults extends BuildCommon {
       val (dirs, files) = preserve.filter(_.exists).flatMap(_.***.get).partition(_.isDirectory)
       val mappings = files.zipWithIndex map { case (f, i) => (f, new File(temp, i.toHexString)) }
       IO.move(mappings)
-      IO.delete(clean)
+      clean.foreach(root => IO.delete((root.*** --- root).get))
```

Comment:
One way to mitigate potential perf regression might be to avoid `***` here and scan only the first level of items using `*`.